tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs

```diff
@@ -39,43 +41,36 @@ public Task<bool> Ping()
         public void WriteTrace(ArraySegment<Span> trace)
         {
             // We ensure there's no trace (local root span) without a test tag.
-            // And ensure all remaining spans have the origin tag.
+            // And ensure all other spans have the origin tag.
 
-            HashSet<ulong> removeIds = null;
-            Span[] finalTrace = new Span[trace.Count];
-            int idx = 0;
-            foreach (var span in trace)
+            // Check if the trace has any span
+            if (trace.Count == 0)
             {
-                // Remove traces non test, benchmarks or build traces.
-                if (span.Context.Parent is null)
-                {
-                    if (span.Type != SpanTypes.Test &&
-                        span.Type != SpanTypes.Benchmark &&
-                        span.Type != SpanTypes.Build)
-                    {
-                        if (removeIds == null)
-                        {
-                            removeIds = new HashSet<ulong>();
-                        }
+                // No trace to write
+                return;
+            }
 
-                        removeIds.Add(span.SpanId);
-                        CIVisibility.Log.Warning("Non Test or Benchmark trace was dropped: {Span}", span);
-                        continue;
-                    }
-                }
-                else if (removeIds != null && removeIds.Contains(span.Context.ParentId.Value))
+            if (!_isPartialFlushEnabled)
+            {
+                // Check if the last span (the root) is a test, bechmark or build span
+                Span lastSpan = trace.Array[trace.Offset + trace.Count - 1];
+                if (lastSpan.Context.Parent is null &&
+                    lastSpan.Type != SpanTypes.Test &&
+                    lastSpan.Type != SpanTypes.Benchmark &&
+                    lastSpan.Type != SpanTypes.Build)
                 {
-                    removeIds.Add(span.SpanId);
-                    CIVisibility.Log.Warning("Non Test or Benchmark trace was dropped: {Span}", span);
-                    continue;
+                    CIVisibility.Log.Warning<int>("Spans dropped because not having a test or benchmark root span: {Count}", trace.Count);
+                    return;
                 }
+            }
```

Comment:
This is undoubtedly simpler if partial flush isn't enabled, but do we still need your previous implementation if it _is_ enabled? 

Or, if we don't support partial flush enabled, should we not write _any_ traces, as with this implementation, I think you could still be sending traces that don't have a test or benchmark root

Reply:
Not sending the trace is really an optimisation because those traces will be invisible in the UI. For billing purposes the important part is the origin tag we are already doing to all spans.

If we decide to do the optimisation for partial flushing later, I prefer to do it in another PR.